### PR TITLE
Resolve concrete associated indices when resetting features

### LINKED ISSUE #72363: Make feature reset compatible with action.destructive_requires_name: false
It has long been best practice to set `action.destructive_requires_name: false` and that is becoming the default in 8.0.

Currently if an end user runs the feature reset API they are likely to see this error response:

```
      "feature_name" : "machine_learning",
      "status" : "FAILURE",
      "exception" : {
        "type" : "exception",
        "reason" : "java.lang.IllegalArgumentException: Wildcard expressions or all indices are not allowed",
        "caused_by" : {
          "type" : "illegal_argument_exception",
          "reason" : "Wildcard expressions or all indices are not allowed"
        }
      }
    },
```

This problem occurs because the feature reset API is trying to delete wildcarded index patterns.  The solution is to set `action.destructive_requires_name: true`, which is what our integration test suites do to avoid this problem, but is not great UX for an end user (and the error response does not make the solution clear). 

It would be much friendlier to resolve the wildcard patterns to a list of concrete indices as suggested in point 2 of https://github.com/elastic/elasticsearch/issues/71178#issuecomment-812193281, or else bypass the value of `action.destructive_requires_name` in some other way within the internals of the feature reset API implementation.

### PR DESCRIPTION
**Pull request by williamrandolph**
When associated index patterns contained wildcards and `action.destructive_requires_name` was set to `true`, feature resets were failing. In order to avoid this, we want to resolve associated index names, then pass the concrete index names to the delete request.

For normal system indices, the `SystemIndexDescriptor` provides a method that searches cluster metadata for indices that match the the system index pattern. This PR introduces a more general index descriptor class that provides the same mechanism. Although we could find a way to use an `IndexNameExpressionResolver` for the associated indices, it makes sense to me to keep things consistent across the various feature index pattern collections.

This change has the effect of allowing the same regex-like syntax that system index patterns can use, rather than just wildcards, in associated index patterns.

I have tried to eliminate duplicated code by using a class hierarchy here, but if we prefer to avoid this layer of inheritance, I can roll back that commit without any trouble and simply use an `AssociatedIndexDescriptor` class.

Fixes #72363

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.lucene.util.automaton.RegExp;
+import org.elasticsearch.cluster.metadata.Metadata;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * An "associated index" is an index that is related to or derived from a system
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java:22]: If we want to keep the `IndexDescriptor` superclass, I will need to rewrite this javadoc.

File: server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    private final CharacterRunAutomaton indexPatternAutomaton;
+
+    /**
+     * Create a descriptor for an index associated with a feature
+     * @param indexPattern Pattern for associated indices, which can include wildcards
+     *                     or regex-like character classes
+     * @param description A short description of the purpose of this index
+     */
+    public IndexDescriptor(String indexPattern, String description) {
+        Objects.requireNonNull(indexPattern, "associated index pattern must not be null");
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java:47]: These error messages should say `system index` if a system index throws the error.

File: server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+/**
+ * An "associated index" is an index that is related to or derived from a system
+ * index, but should not be considered a system index, usually because it is
+ * meant to be visible to users. However, if it goes out of sync with its
+ * related system indices, it may become unreliable or useless. Hence, when
+ * taking a snapshot of a feature, we want the associated index to be included
+ * in the snapshot, and, likewise, we want an associated index to be restored
+ * when a feature snapshot is restored.
+ */
+public class IndexDescriptor {
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java:30]: make it abstract?

File: server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        return Collections.unmodifiableList(matchingIndices);
+    }
+
+    /**
+     * Checks whether an index name matches the system index name pattern for this descriptor.
+     * @param index The index name to be checked against the index pattern given at construction time.
+     * @return True if the name matches the pattern, false otherwise.
+     */
+    public boolean matchesIndexPattern(String index) {
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java:116]: make this method `final`?

**Review by jaymode — COMMENTED** [review]
I think the base class is fine. I left a few small comments.

File: server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.elasticsearch.indices;
+
+import org.elasticsearch.cluster.metadata.Metadata;
+
+import java.util.List;
+
+/**
+ * An IndexDescriptor is an object that
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java:16]: ...that does what? I probably should have finished this sentence.

- [ ] finish writing this javadoc

**Review by jaymode — APPROVED** [review]
LGTM

**Comment by williamrandolph** [comment]
Backport commits:

7.x: cb386ecf87a51e643b5f009aebde1ca5dfcaea9b
7.13: 0479e24b75d

**Pull request merged**